q2_longitudinal/tests/test_longitudinal.py

```diff
@@ -288,6 +288,33 @@ def test_nmit_missing_table_ids(self):
             nmit(table=self.table_taxa_fp, metadata=md,
                  individual_id_column='studyid')
 
+    def test_first_differences(self):
```

Comment:
Thanks @jairideout ! I have added a whole bunch of tests. Many of these issues (e.g., column validation) are checked by utility functions that already existed, but I have added tests for some of these anyway. I have also added several more validator functions in response to your points — you bring up a number of very good cases!

> If table is provided, that behavior is probably okay (but worth having a unit test for)
Why do you think it would be okay for metric to match other columns if the metric is in a table? I can make workarounds for that case, but for now I've added a validator to make sure *no columns* match because as far as I can think this may lead to unintended consequences.

> Can you add another valid test data set
I have one real data set and one toy data set. What kind of additional datasets would you want to see (real or toy)?

Reply:
> Why do you think it would be okay for metric to match other columns if the metric is in a table? I can make workarounds for that case, but for now I've added a validator to make sure no columns match because as far as I can think this may lead to unintended consequences.

I was thinking it'd be technically acceptable to supply a `metric` in the feature table which could overlap with the same column name in `metadata` used for `individual_id_column` or `state_column`. We know it isn't the same column being supplied because they're coming from different files (`table` vs `metadata`). For example, imagine your `metric` in the feature table is called `foo`, and you also happen to have an `individual_id_column` called `foo`. Those `foo` names refer to columns in different files so it *could* be a valid operation, but also a very odd border case.

I think this is such a rare/odd border case, and the behavior wouldn't be clear in that situation unless the user carefully read the method's docs. The validator behavior you described is better in the long run, let's go with that!

> I have one real data set and one toy data set already implemented. What kind of additional datasets would you want to see added (real or toy)?

Apologies, I thought there was only one "valid" data set being tested, I missed the other one. I think with the addition of the data sets necessary to test those border cases I listed above, we should be GTG!

Reply:
Thanks for adding these test cases! :+1: 

Can you also add a test case where the sequential states aren't in any kind of order in the metadata file (i.e. to test that they are sorted correctly within the method)? It'd also be good to have a test case where the order of sample IDs in the metadata file don't match the order of sample IDs in the table or distance matrix. Apologies if you already have these test cases, from a quick look I didn't see it.